Add story pinned option

griseduardo opened the pull request:
Motivation
------------------------------------
Allow pin story and show first

Changes
----------------------------------

- Add story_pinned column in stories table
- Allow editor and admin user to pin or unpin stories in stories dashboard and edit story dashboard
- When a story is pinned, any other previous pinned is unpinned
- Member user can only see the pin in pinned story but not update it
- In home and dashboard show the pinned story first (if filters satisfy the condition to this story appear)

Screenshots
----------------------------------
 Stories dashboard (admin/editor):
![stories-dashboard-admin-editor](https://github.com/Terrastories/terrastories/assets/34499486/8b77d284-82ed-4683-b153-38cbbf39e0f3)

Stories dashboard (member):
![stories-dashboard-member](https://github.com/Terrastories/terrastories/assets/34499486/b888ec9c-0adf-4360-95ef-fbb3cebc26f1)

Story dashboard (admin/editor):
![story-dashboard-admin-editor](https://github.com/Terrastories/terrastories/assets/34499486/b76cb1a9-d2f7-4519-aa70-354be3b49a04)

Home page:
![home](https://github.com/Terrastories/terrastories/assets/34499486/6d227ba4-174d-479e-b07a-de2d0a6e3860)

rudokemper commented:
This is great to see! Thanks so much.

A couple of things:

1. The addition of these css assets are introducing a lot of unwanted styling artifacts on both the Rails dashboard and React sides. 

```
    <link href="//netdna.bootstrapcdn.com/twitter-bootstrap/2.3.2/css/bootstrap-combined.no-icons.min.css" rel="stylesheet">
    <link href="//netdna.bootstrapcdn.com/font-awesome/3.2.1/css/font-awesome.css" rel="stylesheet">
 ```

You can see this in the blue underlines in your GIFs above, as well as the the New button at the top right which is supposed to be orange background, and a bunch of other little things. And on React, the light blue link color at the bottom. 

Can we find a better way to access the icons without importing a bunch of CSS we either don't need or is impacting the app style?

2. For a story that's been "Pinned" can we add a permanently visible pin icon at the top right?

rudokemper commented on rails/app/views/layouts/dashboard.html.erb:
```diff
@@ -1,6 +1,8 @@
 <!DOCTYPE html>
 <html lang="<%= I18n.locale %>">
   <head>
+    <link href="//netdna.bootstrapcdn.com/twitter-bootstrap/2.3.2/css/bootstrap-combined.no-icons.min.css" rel="stylesheet">
```
See comment above.

griseduardo commented:
> This is great to see! Thanks so much.
> 
> A couple of things:
> 
>     1. The addition of these css assets are introducing a lot of unwanted styling artifacts on both the Rails dashboard and React sides.
> 
> 
> ```
>     <link href="//netdna.bootstrapcdn.com/twitter-bootstrap/2.3.2/css/bootstrap-combined.no-icons.min.css" rel="stylesheet">
>     <link href="//netdna.bootstrapcdn.com/font-awesome/3.2.1/css/font-awesome.css" rel="stylesheet">
> ```
> 
> You can see this in the blue underlines in your GIFs above, as well as the the New button at the top right which is supposed to be orange background, and a bunch of other little things. And on React, the light blue link color at the bottom.
> 
> Can we find a better way to access the icons without importing a bunch of CSS we either don't need or is impacting the app style?
> 
>     2. For a story that's been "Pinned" can we add a permanently visible pin icon at the top right?

Hello, the problem is that the exact "icon-pushpin" is only presented in version 3 of font-awesome: https://fontawesome.com/v3/icons/
It is a version that is no longer supported and it is only possible to use with bootstrap (like I tried before) or downloading a folder with font-awesome configs, that brings all icons from there (not separated by icon).
The icons presented in version 5/6 of font-awesome, they permit download the svg of free ones and they give support for them.
I downloaded the similar of pushpin of version 6 and used it, I updated the gifs in description to show how the icon appears. 
What do you think about use this svg?
I updated to always show pin icon when the story is pinned

julinvictus commented:
@griseduardo You think you can rebase your branch so we dont have the conflicts?

lauramosher commented on rails/app/assets/images/thumbtack-not-pinned.svg:
```diff
@@ -0,0 +1 @@
+<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 384 512"><!--! Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license (Commercial License) Copyright 2023 Fonticons, Inc. --><style>svg{fill:#a6a6a6}</style><path d="M32 32C32 14.3 46.3 0 64 0H320c17.7 0 32 14.3 32 32s-14.3 32-32 32H290.5l11.4 148.2c36.7 19.9 65.7 53.2 79.5 94.7l1 3c3.3 9.8 1 ...
```
I know the issue stated to use "font-awesome" — which is our bad on leaving that in when we've since removed font-awesome 😅. I'm wondering if we would benefit from switching to using icons from Google's open source [Material Icons](https://fonts.google.com/icons?selected=Material%20Symbols%20Outlined%3Apush_pin%3AFILL%401%3Bwght%40400%3BGRAD%400%3Bopsz%4048) which has supplied several of our other icons in recent icon additions.

This isn't blocking.

griseduardo replied:
No problem, I updated to use google icon svg

lauramosher reviewed (changes requested):
Overall, this a great start! I do have a few concerns that I've outlined below.

I would also love to see tests added surrounding the actions for pinning and unpinning. While unlikely, it isn't impossible for pinned stories to get out of sync and result in a community having multiple stories pinned. With the way the code is right now, if that were to happen, we would only ever unpin the first found pinned story rather than all of them.

Thank you for your contribution and I apologize for the rather tardy review.

griseduardo commented on rails/app/pages/stories_page.rb:
```diff
@@ -5,7 +5,8 @@ def initialize(scoped_stories, meta = {})
     @meta[:limit] ||= 10
     @meta[:offset] ||= 0
-    @meta[:sort_by] ||= "created_at"
+    @meta[:sort_by_pinned] ||= "story_pinned"
+    @meta[:sort_by_created] ||= "created_at"
```
It is true. I will return the previous behavior

griseduardo commented on rails/app/pages/stories_page.rb:
```diff
@@ -16,6 +17,7 @@ def relation
     stories = stories.where(permission_level: @meta[:visibility]) if @meta[:visibility].present?
 
-    stories.order(@meta[:sort_by] => @meta[:sort_dir])
+    stories = stories.order(@meta[:sort_by_pinned] => @meta[:sort_dir])
```
Nice, I combined the order rules

griseduardo commented on rails/app/assets/stylesheets/cms/_cards.scss:
```diff
@@ -117,3 +117,29 @@
+}
+
+.edit-icon-pinned, .view-icon-pinned {
+  color: $black;
```
It was filling the SVG color in previous behavior, but now doesn't need anymore. I removed it

griseduardo commented on rails/app/assets/stylesheets/cms/_cards.scss:
```diff
@@ -117,3 +117,29 @@
+  float: right;
+  height: fit-content;
+  margin-inline-start: auto;
+}
```
It is true, the change worked here

griseduardo commented on rails/app/assets/stylesheets/cms/_cards.scss:
```diff
@@ -117,3 +117,29 @@
+
+.card:hover .edit-icon-not-pinned {
+  display: flex;
+}
```
Good idea, it worked here

griseduardo commented on rails/app/controllers/dashboard/stories_controller.rb:
```diff
@@ -43,8 +43,29 @@ def edit
+
+        @story.update(story_pinned: story_params[:story_pinned])
+
+        render :edit
```
Good idea, I updated here

griseduardo commented on rails/app/assets/stylesheets/cms/_icons.scss:
```diff
@@ -28,4 +28,15 @@
+
+.icon-not-pinned {
+  color: $light-gray;
+}
```
I removed css rules from _icons.scss

griseduardo commented on rails/app/assets/stylesheets/cms/layout.scss:
```diff
@@ -172,3 +172,24 @@ svg[hidden] {
+.image-pinned {
+  height: 15px;
+  width: 9px;
+}
```
Changing to inside edit form, those css rules are no longer necessary. I removed here

griseduardo commented on rails/app/controllers/dashboard/stories_controller.rb:
```diff
@@ -43,8 +43,29 @@ def edit
+        @story.update(story_pinned: params[:story_pinned])
+
+        return redirect_to action: "index"
+      end
```
Good idea separate in different actions.
Only one problem, related to use remote true, is that it doesn't change the icon color and reordenize the stories, that can pass the wrong idea that the pin or unpin action didn't worked. 
I maintained the render behavior. What do you think about it?

griseduardo commented on rails/app/assets/stylesheets/cms/_cards.scss:
```diff
@@ -117,3 +117,29 @@
+  &:hover {
+    box-shadow: 0 0 10px $lighter-gray;
+  }
+}
```
Awesome. I just updated a bit because this way the color only changed when the mouse is on icon, but on button doesn't change the color of icon.
I updated to change the color of icon when the mouse is on the button

griseduardo commented on rails/app/views/dashboard/stories/_stories.html.erb:
```diff
@@ -1,10 +1,40 @@
+              <% end %>
+            </div>
+          <% end %>
+        <% end %>
```
Nice idea, I updated to use the svg this way in rails. Anyway I tried and doesn't find for now a way to use the same in React. I used the svg directly there. Do you have a suggestion for how to make it work?

lauramosher commented on rails/app/views/dashboard/stories/_stories.html.erb:
```diff
@@ -1,10 +1,38 @@
+          </div>
+        <% end %>
+        <% if current_user.admin? || current_user.editor? %>
+          <% if story.story_pinned %>
+            <div class="edit-icon-pinned">
+              <%= form_with model: @story, url: unpin_story_path(story), method: "patch", multipart: true, class: "form", data: {remote: false}, locale: true do |f| %>
+                <%= f.button nil, class: "icon-button" do %>
+                  <svg class="icon-pinned"><use href="#icon-thumbtack"></svg>
+                <% end %>
+              <% end %>
+            </div>
+          <% else %>
+            <div class="edit-icon-unpinned">
+              <%= form_with model: @story, url: pin_story_path(story), method: "patch", multipart: true, class: "form", data: {remote: false}, locale: true do |f| %>
+                <%= f.button nil, class: "icon-button" do %>
+                  <svg class="icon-unpinned"><use href="#icon-thumbtack"></svg>
+                <% end %>
+              <% end %>
+            </div>
+          <% end %>
```
I think this could benefit from using Rails' [`button_to` helper](https://api.rubyonrails.org/v7.1.1/classes/ActionView/Helpers/UrlHelper.html#method-i-button_to). It essentially creates the code you're doing here, but streamlines the form creation.

```suggestion
          <% if story.story_pinned %>
            <%= button_to nil, unpin_story_path(story), method: "patch", class: "edit-icon-pinned", data: {remote: false}, locale: true  do %>
              <svg class="icon-pinned"><use href="#icon-thumbtack"></svg>
            <% end %>
          <% else %>
            <%= button_to nil, pin_story_path(story), method: "patch", class: "edit-icon-unpinned", data: {remote: false}, locale: true do %>
              <svg class="icon-unpinned"><use href="#icon-thumbtack"></svg>
            <% end %>
          <% end %>
```

griseduardo replied:
NIce idea, I updated here

lauramosher reviewed (commented):
Thanks for working through all of the feedback! This is looking great. A couple of additional comments in line.

I would also love to see some specs around the pinning and unpinning actions — both the individual pin/unpin endpoints as well as in the create and edit form. Are you able to add those for us?

griseduardo commented:
> Thanks for working through all of the feedback! This is looking great. A couple of additional comments in line.
> 
> I would also love to see some specs around the pinning and unpinning actions — both the individual pin/unpin endpoints as well as in the create and edit form. Are you able to add those for us?

I will add the tests :)
I don't download Docker in my machine, so it take some time to adjust the code to run docker-compose without it, and now to make the already presented tests work too. But I figureted how to make Rails tests run, and I will add the new tests

griseduardo commented on rails/app/assets/stylesheets/cms/_cards.scss:
```diff
@@ -117,3 +126,58 @@
+
+.icon-pinned {
+  fill: $black;
+}
```
Good idea, I updated here

griseduardo commented on rails/app/assets/stylesheets/cms/_cards.scss:
```diff
@@ -29,6 +29,15 @@
+
+  &--header {
+    display: flex;
+    justify-content: space-between;
```
I removed and didn't change the layout behavior, my mistake :sweat_smile:

griseduardo commented:
Hello @rudokemper and @lauramosher 
I believe that now is ready to re-review, I worked in all suggestions.
The resolved ones I resolve here in pr, some I commented the implementation I followed and the reason :)
